java-frontend/src/main/java/org/sonar/java/se/ExplodedGraphWalker.java

```diff
@@ -582,39 +587,111 @@ private void executeMethodInvocation(MethodInvocationTree mit) {
+        invocationArguments,
+        invocationTypes,
+        programState,
+        () -> thrownExceptionsByExceptionType.computeIfAbsent(yield.exceptionType, constraintManager::createExceptionalSymbolicValue))
+      ).forEach(ps -> enqueueExceptionalPaths(ps.clearStack(), (SymbolicValue.ExceptionalSymbolicValue) ps.peekValue()));
 
+      // Enqueue additional exceptional paths corresponding to unchecked exceptions, for instance OutOfMemoryError
+      enqueueUncheckedExceptionalPaths();
+
+      // Enqueue happy paths
       methodInvokedBehavior.happyPathYields()
-        .flatMap(yield -> yield.statesAfterInvocation(invocationArguments, invocationTypes, programState, () -> resultValue).stream())
+        .flatMap(yield -> yield.statesAfterInvocation(invocationArguments, invocationTypes, programState, () -> resultValue))
+        .map(psYield -> handleSpecialMethods(psYield, mit, methodSymbol))
         .forEach(psYield -> {
-          ProgramState ps = psYield;
-          if (isNonNullMethod(methodSymbol)) {
-            ps = ps.addConstraint(ps.peekValue(), ObjectConstraint.NOT_NULL);
-          } else if (OBJECT_WAIT_MATCHER.matches(mit)) {
-            ps = ps.resetFieldValues(constraintManager);
-          }
           checkerDispatcher.syntaxNode = mit;
-          checkerDispatcher.addTransition(ps);
+          checkerDispatcher.addTransition(psYield);
           clearStack(mit);
         });
     } else {
-      enqueueExceptionalPaths(programState);
+      // Enqueue exceptional paths from thrown exceptions
+      enqueueThrownExceptionalPaths(methodSymbol);
 
-      programState = programState.stackValue(resultValue);
-      if (isNonNullMethod(methodSymbol)) {
-        programState = programState.addConstraint(resultValue, ObjectConstraint.NOT_NULL);
-      } else if (OBJECT_WAIT_MATCHER.matches(mit)) {
-        programState = programState.resetFieldValues(constraintManager);
-      }
+      // Enqueue additional exceptional paths corresponding to unchecked exceptions, for instance OutOfMemoryError
+      enqueueUncheckedExceptionalPaths();
+
+      // Enqueue happy paths
+      programState = handleSpecialMethods(programState.stackValue(resultValue), mit, methodSymbol);
       checkerDispatcher.executeCheckPostStatement(mit);
       clearStack(mit);
     }
   }
 
-  private void enqueueExceptionalPaths(ProgramState ps) {
-    node.programPoint.block.exceptions().forEach(b -> enqueue(new ExplodedGraph.ProgramPoint(b, 0), ps.clearStack(), !b.isCatchBlock()));
+  private ProgramState handleSpecialMethods(ProgramState ps, MethodInvocationTree mit, Symbol methodSymbol) {
+    if (isNonNullMethod(methodSymbol)) {
+      return ps.addConstraint(ps.peekValue(), ObjectConstraint.NOT_NULL);
+    } else if (OBJECT_WAIT_MATCHER.matches(mit)) {
+      return ps.resetFieldValues(constraintManager);
+    }
+    return ps;
+  }
+
+  private void enqueueThrownExceptionalPaths(Symbol symbol) {
```

Comment:
I think this one could rely on MethodInvocationTree : that would allow to not declare methodSymbol local variable.